### PR TITLE
Feature/dds publisher

### PR DESCRIPTION
**Pull request by IkerLuengo**
Due to cross dependencies, this PR is developed on a branch originating from that of #114 (Domain module documentation) and includes all changes from #114. Specifically all commits referring to task 8330 are from #114 while the ones referring to task 8423 are related to the Publisher module.

This PR should be merged after #114.

### CODE REVIEW

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/307/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/309/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/310/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/311/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/314/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/320/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/321/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/322/)

File: docs/fastdds/dds_layer/publisher/publisher/publisher.rst [hunk 1]
```diff
@@ -3,3 +3,160 @@
...
+.. _dds_layer_publisher_publisherQos:
+
+PublisherQos
+------------
+
+:class:`PublisherQos` controls the behavior of the :ref:`dds_layer_publisher_publisher`.
+Internally it contains the following :class:`QosPolicy` objects:
+
++--------------------------------+------------------------------------+----------+
+| QosPolicy class                | Accessor                           | Mutable  |
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisher/publisher.rst:22]:
```suggestion
| QosPolicy class                | Accessor/Mutator                   | Mutable  |
```

File: docs/fastdds/dds_layer/publisher/publisher/publisher.rst [hunk 2]
```diff
@@ -3,3 +3,160 @@
...
+| EntityFactoryQosPolicy         | :func:`entity_factory`             | Yes      |
++--------------------------------+------------------------------------+----------+
+
+Refer to the detailed description of each :class:`QosPolicy` class for more information about their usage and
+default values.
+
+The QoS value of a previously created :ref:`dds_layer_publisher_publisher` can be modified using the
+:func:`set_qos` member function.
+
+.. literalinclude:: /../code/DDSCodeTester.cpp
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisher/publisher.rst:39]: In general, indentation

File: docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst [hunk 3]
```diff
@@ -3,3 +3,20 @@
...
+
+:class:`PublisherListener` inherits from :ref:`dds_layer_publisher_dataWriterListener`.
+Therefore, it has the ability to react to all events that are reported to the
+:ref:`dds_layer_publisher_dataWriter`.
+Since events are always notified to the most specific Entity Listener that can handle the event,
+callbacks that :class:`PublisherListener` inherits from :ref:`dds_layer_publisher_dataWriterListener`
+will only be called if the triggering :ref:`dds_layer_publisher_dataWriter` has
+no Listener attached.
+
+:class:`PublisherListener` does not add any new callback.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst:21]: Maybe a link to the DataWriterListener section, although I'd love a snippet here

File: docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst [hunk 4]
```diff
@@ -3,3 +3,261 @@
...
+.. _dds_layer_publisher_dataWriterQos:
+
+DataWriterQos
+-------------
+
+:class:`DataWriterQos` controls the behavior of the :ref:`dds_layer_publisher_dataWriter`.
+Internally it contains the following :class:`QosPolicy` objects:
+
++--------------------------------+------------------------------------+----------+
+| QosPolicy class                | Accessor                           | Mutable  |
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst:28]:
```suggestion
| QosPolicy class                | Accessor/Mutator                   | Mutable  |
```

File: code/DDSCodeTester.cpp [hunk 5]
```diff
@@ -0,0 +1,1041 @@
...
+        DomainParticipant* participant =
+                DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
+        if (nullptr != participant)
+        {
+            // Error
+            return;
+        }
+
+        // Create a Publisher with default PublisherQos and no Listener
+        // The symbol PUBLISHER_QOS_DEFAULT is used to denote the default QoS.
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:466]: In general, value

File: code/DDSCodeTester.cpp [hunk 6]
```diff
@@ -0,0 +1,1041 @@
...
+    Publisher* publisher =
+            participant->create_publisher(PUBLISHER_QOS_DEFAULT);
+    Topic* topic =
+            participant->create_topic("TopicName", "DataTypeName", TOPIC_QOS_DEFAULT);
+
+    {
+        //DDS_CREATE_DATAWRITER
+        // Create a DataWriter with default DataWriterQos and no Listener
+        // The symbol DATAWRITER_QOS_DEFAULT is used to denote the default QoS.
+        DataWriter* dataWriter_with_default_qos =
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:794]: I wouldn't mix camel and snake-case. Fast DDS style guide goes with snake, so I'd stick to that

File: docs/fastdds/dds_layer/publisher/dataWriterListener/dataWriterListener.rst [hunk 7]
```diff
@@ -3,3 +3,40 @@
...
+   data within the deadline period configured on its :ref:`dds_layer_publisher_dataWriterQos`.
+   It will be called for each deadline period and data instance for which the
+   :ref:`dds_layer_publisher_dataWriter` failed to provide data.
+
+ * **on_offered_incompatible_qos**: The :ref:`dds_layer_publisher_dataWriter` has found a
+   :ref:`dds_layer_subscriber_dataReader` that matches the :ref:`dds_layer_topic_topic` and has
+   a common partition, but with a requested QoS that is incompatible with the one defined on the
+   :ref:`dds_layer_publisher_dataWriter`.
+
+   Currently this callback is not implemented (it will never be called), and will be implemented
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriterListener/dataWriterListener.rst:30]: In the HTML it's not clean whether the one not implemented is `on_offered_incompatible_qos` or `on_liveliness_lost`

File: code/DDSCodeTester.cpp [hunk 8]
```diff
@@ -0,0 +1,1041 @@
...
+        (void)writer, (void)status;
+        std::cout << "Found a remote Topic with incompatible QoS" << std::endl;
+    }
+
+    /**
+    * @brief Method called when the livelivess of a publisher is lost
+    * @param writer The publisher
+    * @param status The liveliness lost status
+    */
+    virtual void on_liveliness_lost(
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:770]: This is the only callback with doxygen doc

File: docs/fastdds/dds_layer/publisher/publisher/publisher.rst [hunk 9]
```diff
@@ -3,3 +3,160 @@
...
+   :language: c++
+   :start-after: //DDS_CHANGE_PUBLISHERQOS_TO_DEFAULT
+   :end-before: //!
+
+
+.. _dds_layer_publisher_creation:
+
+Creating a Publisher
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisher/publisher.rst:81]: I'd put this and `Creating a DataWriter` as standalone pages. Same spot in the toctree and all, but just by themselves

File: docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst (hunk 3 above)
Comment (IkerLuengo) [reply docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst:21]: The link is in the previous paragraph

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/330/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/333/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/334/)

File: docs/fastdds/dds_layer/publisher/publisher/publisher.rst [hunk 10]
```diff
@@ -3,3 +3,84 @@
...
+
+.. _dds_layer_defaultPublisherQos:
+
+Default PublisherQos
+^^^^^^^^^^^^^^^^^^^^
+
+The default :ref:`dds_layer_publisher_publisherQos` refers to the value returned by the
+:func:`get_default_publisher_qos` member function on the :ref:`dds_layer_domainParticipant` instance.
+The special value ``PUBLISHER_QOS_DEFAULT`` can be used as QoS argument on :func:`create_publisher`
+or :func:`set_qos` member functions to indicate that the current default :ref:`dds_layer_publisher_publisherQos`
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisher/publisher.rst:59]: Is it to use the current default or the constructed default? Below it says constructed

File: docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst (hunk 3 above)
Comment (EduPonz) [reply docs/fastdds/dds_layer/publisher/publisherListener/publisherListener.rst:21]: You're right, but I'd still point the user there with a "look here" kind of thing

File: docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst [hunk 11]
```diff
@@ -3,3 +3,128 @@
...
+
+.. _dds_layer_defaultDataWriterQos:
+
+Default DataWriterQos
+^^^^^^^^^^^^^^^^^^^^^
+
+The default :ref:`dds_layer_publisher_dataWriterQos` refers to the value returned by the
+:func:`get_default_dataWriter_qos` member function on the :ref:`dds_layer_publisher_publisher` instance.
+The special value ``DATAWRITER_QOS_DEFAULT`` can be used as QoS argument on :func:`create_datawriter`
+or :func:`set_qos` member functions to indicate that the current default :ref:`dds_layer_publisher_dataWriterQos`
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst:101]: Same as before, is it current or constructed?

File: docs/fastdds/dds_layer/publisher/dataWriterListener/dataWriterListener.rst [hunk 12]
```diff
@@ -3,3 +3,41 @@
...
+  :ref:`dds_layer_publisher_dataWriter` failed to provide data.
+
+* **on_offered_incompatible_qos**: The :ref:`dds_layer_publisher_dataWriter` has found a
+  :ref:`dds_layer_subscriber_dataReader` that matches the :ref:`dds_layer_topic_topic` and has
+  a common partition, but with a requested QoS that is incompatible with the one defined on the
+  :ref:`dds_layer_publisher_dataWriter`.
+
+.. note::
+   Currently *on_offered_incompatible_qos* is not implemented (it will never be called), and will be implemented
+   on a future release of *fast DDS*.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriterListener/dataWriterListener.rst:32]:
```suggestion
   on a future release of *Fast DDS*.
```

File: docs/fastdds/dds_layer/publisher/dataWriter/publishingData.rst [hunk 13]
```diff
@@ -0,0 +1,49 @@
+.. _dds_layer_publisher_write:
+
+Publishing data
+===============
+
+The user informs of a change in the value of a data instance with the :func:`write` member function on the
+:ref:`dds_layer_publisher_dataWriter`. This change will then be communicated to every
+:ref:`dds_layer_subscriber_subscriber` matched with the :ref:`dds_layer_publisher_dataWriter`.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriter/publishingData.rst:8]: It's the `DataReader` the one that matches with the `DataWriter`

File: docs/fastdds/dds_layer/domain/domain.rst [hunk 14]
```diff
@@ -3,6 +3,8 @@
 Domain
 ======
 
+.. image:: /01-figures/domain_class_diagram.svg
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/domain/domain.rst:6]: I believe this is a mistake from the rebase

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/338/)

File: docs/fastdds/dds_layer/domain/domainParticipant/domainParticipant.rst [hunk 15]
```diff
@@ -94,4 +94,9 @@ This will reset the current default :ref:`dds_layer_domainParticipantQos` to the
    :end-before: //!
    :dedent: 8
 
+.. note::
+   The value ``PARTICIPANT_QOS_DEFAULT`` has different meaning depending on where it is used:
+   * On :func:`create_participant` and :func:`set_qos` it refers to the default :ref:`dds_layer_domainParticipantQos`
+     as returned by :func:`get_default_participant_qos`.
+   * On :func:`set_default_participant_qos` it refers to the default constructed :func:`DomainParticipantQos`.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/domain/domainParticipant/domainParticipant.rst:101]:
```suggestion
.. note::
   The value ``PARTICIPANT_QOS_DEFAULT`` has different meaning depending on where it is used:

   * On :func:`create_participant` and :func:`set_qos` it refers to the default :ref:`dds_layer_domainParticipantQos`
     as returned by :func:`get_default_participant_qos`.
   * On :func:`set_default_participant_qos` it refers to the default constructed :func:`DomainParticipantQos`.
```

File: docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst [hunk 16]
```diff
@@ -3,3 +3,128 @@
...
+This will reset the current default :ref:`dds_layer_publisher_dataWriterQos` to default constructed
+value :func:`DataWriterQos`.
+
+.. literalinclude:: /../code/DDSCodeTester.cpp
+   :language: c++
+   :start-after: //DDS_CHANGE_DATAWRITERQOS_TO_DEFAULT
+   :end-before: //!
+   :dedent: 8
+
+
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/dataWriter/dataWriter.rst:128]: Add note about default

File: docs/fastdds/dds_layer/publisher/publisher/publisher.rst [hunk 17]
```diff
@@ -3,3 +3,89 @@
...
+   :language: c++
+   :start-after: //DDS_CHANGE_PUBLISHERQOS_TO_DEFAULT
+   :end-before: //!
+   :dedent: 8
+
+.. note::
+   The value ``PUBLISHER_QOS_DEFAULT`` has different meaning depending on where it is used:
+   * On :func:`create_publisher` and :func:`set_qos` it refers to the default :ref:`dds_layer_publisher_publisherQos`
+     as returned by :func:`get_default_publisher_qos`.
+   * On :func:`set_default_publisher_qos` it refers to the default constructed :func:`PublisherQos`.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/publisher/publisher/publisher.rst:90]:
```suggestion
.. note::

   The value ``PUBLISHER_QOS_DEFAULT`` has different meaning depending on where it is used:
   * On :func:`create_publisher` and :func:`set_qos` it refers to the default :ref:`dds_layer_publisher_publisherQos`
     as returned by :func:`get_default_publisher_qos`.
   * On :func:`set_default_publisher_qos` it refers to the default constructed :func:`PublisherQos`.
```

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/340/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/341/)

**Pull request merged**